cpp/ql/src/Security/CWE/CWE-022/TaintedPath.ql

```diff
@@ -46,18 +47,44 @@ class FileFunction extends FunctionWithWrappers {
   override predicate interestingArg(int arg) { arg = 0 }
 }
 
-class TaintedPathConfiguration extends TaintTrackingConfiguration {
-  override predicate isSink(Element tainted) {
-    exists(FileFunction fileFunction | fileFunction.outermostWrapperFunctionCall(tainted, _))
+Expr asSourceExpr(DataFlow::Node node) {
+  result in [node.asConvertedExpr(), node.asDefiningArgument()]
+}
+
+Expr asSinkExpr(DataFlow::Node node) {
+  result = node.asConvertedExpr()
+  or
+  result =
+    node.asOperand()
+        .(SideEffectOperand)
+        .getUse()
+        .(ReadSideEffectInstruction)
+        .getArgumentDef()
+        .getUnconvertedResultExpression()
```

Comment:
Any reason why this disjunct get the unconverted expression, while the first disjunct (i.e., the `node.asConvertedExpr()`) gets the potentially _converted_ expression?

Reply:
`node.asConvertedExpr()` removed some of the duplication. I didn't see a similar effect here.

Reply:
That might just have been because tests coverage is not very good though.

Reply:
Actually changing the last line to `.getConvertedResultExpression()` removes all the duplication for the tests, but I also lose the result for https://github.com/github/codeql/blob/165c882be54eeeb412fe5c2a532753799a0fba92/cpp/ql/test/query-tests/Security/CWE/CWE-022/semmle/tests/test.c#L35-L39

Reply:
Hmm. Is that because `fileFunction.outermostWrapperFunctionCall(expr, _)` only holds when `expr` is an unconverted expression in that case?

Reply:
So it seems we need `getUnconvertedResultExpression()`?

Reply:
Either that, or we need to change `isSink` so that it:
- Marks the converted node as the source (so we don't get path duplication), and
- Calls `fileFunction.outermostWrapperFunctionCall` with the converted node's unconverted expression (so that we still find the right sinks).

Reply:
I was thinking of having `asSinkExpr` do:
```ql
result =
  node.asOperand()
      .(SideEffectOperand)
      .getUse()
      .(ReadSideEffectInstruction)
      .getArgumentDef()
      .getConvertedResultExpression()
```
(i.e., have the node be the converted one to remove the path duplication), and then implement `isSink` as something like:

```ql
override predicate isSink(DataFlow::Node node) {
  exists(FileFunction fileFunction |
    fileFunction.outermostWrapperFunctionCall(asSinkExpr(node).getUnconverted(), _)
  )
}
```
(i.e., we still mark the converted expression as a source, but we use the unconverted expression for the call to `outermostWrapperFunctionCall`).

Note that this last change _might_ mean that we get all the result duplications again (since any converted node whose _unconverted_ expression satisfies `outermostWrapperFunctionCall` will be a sink).

Reply:
> Note that this last change might mean that we get all the result duplications again 

Still the same disappearing result 😄

Reply:
:wat: